src/Mirza/SupplyChain/Handlers/UXUtils.hs

```diff
@@ -66,18 +72,36 @@ listEventsPretty  :: (Member context '[HasDB, HasBRClientEnv],
                   => ST.User
                   -> LabelEPCUrn
                   -> AppM context err [PrettyEventResponse]
-listEventsPretty _user lblUrn = do
-  case urn2LabelEPC . getLabelEPCUrn $ lblUrn of
-    Left err -> throwParseError err
-    Right lbl -> do
-      evs <- runDb $ listEventsQuery lbl
-      traverse func evs
-
-
-func :: (Member context '[HasDB, HasBRClientEnv],
-         Member err     '[AsServiceError, AsServantError, AsSqlError, BT.AsBRError])
-      => Ev.Event -> AppM context err PrettyEventResponse
-func ev = do
+listEventsPretty _user = either throwParseError fetchPrettyEvents . urn2LabelEPC . getLabelEPCUrn
+
+fetchPrettyEvents :: (Member context '[HasDB, HasBRClientEnv],
+                    Member err     '[AsServiceError, AsServantError, AsSqlError, BT.AsBRError])
+                  => LabelEPC
+                  -> AppM context err [PrettyEventResponse]
+fetchPrettyEvents lbl = do
+  evs <- runDb $ listEventsQuery lbl
+  currEvs <- traverse eventToPrettyEvent evs
+  (pEvs :: [PrettyEventResponse] ) <- concat <$> traverse getParentEvents evs
+  let allEvs = currEvs <> pEvs
+  pure $ sortOn ( _eventTime . Ev._when . prettyEvent) $ nub allEvs
+
+
+getParentEvents :: (Member context '[HasDB, HasBRClientEnv],
+                    Member err     '[AsServiceError, AsServantError, AsSqlError, BT.AsBRError])
+                => Ev.Event
+                -> AppM context err [PrettyEventResponse]
+getParentEvents ev = do
+  let mParentLabel = getParent . Ev._what $ ev
+  case mParentLabel of
+    Nothing              -> pure []
+    Just (ParentLabel p) -> fetchPrettyEvents (IL p)
```

Comment:
Instead of executing N+1 queries, you might want to consider recursive CTE's to bring back all the rows in one query.